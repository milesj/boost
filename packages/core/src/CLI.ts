/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

// eslint-disable-next-line import/no-extraneous-dependencies
import { Argv } from 'yargs';
import Tool from './Tool';

export default class CLI {
  static registerGlobalOptions(app: Argv, tool: Tool) {
    app
      .option('config', {
        default: '',
        description: tool.msg('app:cliOptionConfig'),
        string: true,
      })
      .option('debug', {
        boolean: true,
        default: false,
        description: tool.msg('app:cliOptionDebug'),
      })
      .option('locale', {
        default: '',
        description: tool.msg('app:cliOptionLocale'),
        string: true,
      })
      .option('output', {
        default: 3,
        description: tool.msg('app:cliOptionOutput'),
        number: true,
      })
      .option('silent', {
        boolean: true,
        default: false,
        description: tool.msg('app:cliOptionSilent'),
      })
      .option('theme', {
        default: 'default',
        description: tool.msg('app:cliOptionTheme'),
        string: true,
      });
  }
}
