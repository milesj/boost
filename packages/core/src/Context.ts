import { isObject } from '@boost/common';

export default class Context {
  /**
   * Create a new instance of the current context and shallow clone all properties.
   */
  clone(...args: any[]): this {
    // @ts-expect-error
    const context = new this.constructor(...args);

    // Copy enumerable properties
    Object.keys(this).forEach(key => {
      const prop = key as keyof this;
      let value: any = this[prop];

      if (Array.isArray(value)) {
        value = [...value];
      } else if (value instanceof Map) {
        value = new Map(value);
      } else if (value instanceof Set) {
        value = new Set(value);
      } else if (value instanceof Date) {
        value = new Date(value.getTime());
      } else if (isObject<object>(value)) {
        // Dont dereference instances, only plain objects
        if (value.constructor === Object) {
          value = { ...value };
        }
      }

      context[prop] = value;
    });

    return context;
  }
}
