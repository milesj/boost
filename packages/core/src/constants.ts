import { Status } from './types';

export const APP_NAME_PATTERN = /^[a-z]{1}[-a-z0-9]+[a-z]{1}$/u;
export const CONFIG_NAME_PATTERN = /^[a-z]{1}[a-zA-Z0-9]+$/u;
export const EVENT_NAME_PATTERN = /^[-a-z.]+$/u;
export const MODULE_NAME_PATTERN = /^(@[-a-z]+\/)?[-a-z]+$/u;
export const PLUGIN_NAME_PATTERN = /^([a-z]+):[-a-z]+$/u;

export const STATUS_PENDING: Status = 'pending';
export const STATUS_RUNNING: Status = 'running';
export const STATUS_SKIPPED: Status = 'skipped';
export const STATUS_PASSED: Status = 'passed';
export const STATUS_FAILED: Status = 'failed';

// Chalk colors
// yellow = app name, module name
// cyan = file
// magenta = plugin name
