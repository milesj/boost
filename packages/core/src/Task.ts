/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Context from './Context';
import {
  STATUS_PENDING,
  STATUS_RUNNING,
  STATUS_SKIPPED,
  STATUS_PASSED,
  STATUS_FAILED,
} from './constants';
import { Status } from './types';
import Emitter from './Emitter';

export type TaskAction<Ctx extends Context> = (
  context: Ctx,
  value: any,
  task: Task<Ctx>,
) => any | Promise<any>;

export interface TaskMetadata {
  depth: number;
  index: number;
  startTime: number;
  stopTime: number;
}

export default class Task<Ctx extends Context> extends Emitter {
  action: TaskAction<Ctx>;

  // @ts-ignore Set after instantiation
  context: Ctx;

  title: string;

  metadata: TaskMetadata = {
    depth: 0,
    index: 0,
    startTime: 0,
    stopTime: 0,
  };

  output: string = '';

  parent: Task<Ctx> | null = null;

  status: Status = STATUS_PENDING;

  statusText: string = '';

  tasks: Task<Ctx>[] = [];

  constructor(title: string, action: TaskAction<Ctx>) {
    super();

    if (!title || typeof title !== 'string') {
      throw new Error('Tasks require a title.');
    }

    if (action !== null && typeof action !== 'function') {
      throw new Error('Tasks require an executable function.');
    }

    this.action = action;
    this.status = STATUS_PENDING;
    this.title = title;
  }

  /**
   * Return true if the task failed when executing.
   */
  hasFailed(): boolean {
    return this.status === STATUS_FAILED;
  }

  /**
   * Return true if the task executed successfully.
   */
  hasPassed(): boolean {
    return this.status === STATUS_PASSED;
  }

  /**
   * Return true if the task has been completed in any form.
   */
  isComplete(): boolean {
    return this.hasPassed() || this.hasFailed() || this.isSkipped();
  }

  /**
   * Return true if the task has not been executed yet.
   */
  isPending(): boolean {
    return this.status === STATUS_PENDING;
  }

  /**
   * Return true if the task is currently running.
   */
  isRunning(): boolean {
    return this.status === STATUS_RUNNING;
  }

  /**
   * Return true if the task was or will be skipped.
   */
  isSkipped(): boolean {
    return this.status === STATUS_SKIPPED;
  }

  /**
   * Run the current task by executing it and performing any before and after processes.
   */
  async run(context: Ctx, value: any): Promise<any> {
    this.setContext(context);
    this.emit('run', [value]);

    if (this.isSkipped()) {
      this.status = STATUS_SKIPPED;
      this.emit('skip', [value]);

      return Promise.resolve(value);
    }

    this.status = STATUS_RUNNING;
    this.metadata.startTime = Date.now();

    try {
      this.output = await this.action(context, value, this);
      this.status = STATUS_PASSED;
      this.metadata.stopTime = Date.now();
      this.emit('pass', [this.output]);
    } catch (error) {
      this.status = STATUS_FAILED;
      this.metadata.stopTime = Date.now();
      this.emit('fail', [error]);

      throw error;
    }

    this.statusText = '';

    return this.output;
  }

  /**
   * Set the context to be passed around.
   */
  setContext(context: Ctx): this {
    this.context = context;

    return this;
  }

  /**
   * Mark a task as skipped if the condition is true.
   */
  skip(condition: boolean = true): this {
    if (condition) {
      this.status = STATUS_SKIPPED;
    }

    return this;
  }
}
