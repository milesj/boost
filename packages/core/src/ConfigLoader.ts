/* eslint-disable no-cond-assign */

import chalk from 'chalk';
import fs from 'fs-extra';
import glob from 'fast-glob';
import path from 'path';
import JSON5 from 'json5';
import camelCase from 'lodash/camelCase';
import mergeWith from 'lodash/mergeWith';
import { Arguments } from 'yargs-parser';
import optimal, { array, bool, instance, number, shape, string, union, object } from 'optimal';
import formatModuleName from './helpers/formatModuleName';
import handleMerge from './helpers/handleMerge';
import isObject from './helpers/isObject';
import isEmptyObject from './helpers/isEmptyObject';
import requireModule from './helpers/requireModule';
import Tool from './Tool';
import { MODULE_NAME_PATTERN, PLUGIN_NAME_PATTERN } from './constants';
import { Debugger, PackageConfig, PluginSetting } from './types';

export interface ConfigObject {
  [key: string]: any;
}

export type ConfigPathOrObject = string | ConfigObject;

export interface ParseOptions {
  errorOnFunction?: boolean;
}

export default class ConfigLoader {
  debug: Debugger;

  package: PackageConfig = { name: '', version: '0.0.0' };

  parsedFiles: { [path: string]: boolean } = {};

  tool: Tool<any>;

  workspaceRoot: string = '';

  constructor(tool: Tool<any>) {
    this.debug = tool.createDebugger('config-loader');
    this.tool = tool;
  }

  /**
   * Find the config using the --config file path option.
   */
  findConfigFromArg(filePath?: string): ConfigPathOrObject | null {
    this.debug.invariant(
      !!filePath,
      'Looking in --config command line option',
      'Found',
      'Not found',
    );

    return filePath ? { extends: [filePath] } : null;
  }

  /**
   * Find the config in the package.json block under the application name.
   */
  findConfigInPackageJSON(pkg: PackageConfig): ConfigPathOrObject | null {
    const configName = this.getConfigName();
    const config = (pkg as any)[configName];

    this.debug.invariant(
      !!config,
      `Looking in package.json under ${chalk.yellow(configName)} property`,
      'Found',
      'Not found',
    );

    if (!config) {
      return null;
    }

    // Extend from a preset if a string
    if (typeof config === 'string') {
      return { extends: [config] };
    }

    return config;
  }

  /**
   * Find the config using local files commonly located in a configs/ folder.
   */
  findConfigInLocalFiles(root: string): ConfigPathOrObject | null {
    const configName = this.getConfigName();
    const relPath = `configs/${configName}.{js,json,json5}`;
    const configPaths = glob
      .sync(relPath, {
        absolute: true,
        cwd: root,
        onlyFiles: true,
      })
      .map(filePath => String(filePath));

    this.debug.invariant(
      configPaths.length === 1,
      `Looking for local config file: ${relPath}`,
      'Found',
      'Not found',
    );

    if (configPaths.length === 1) {
      this.debug('Found %s', chalk.cyan(path.basename(configPaths[0])));

      return configPaths[0];
    }

    if (configPaths.length > 1) {
      throw new Error(this.tool.msg('errors:multipleConfigFiles', { configName }));
    }

    return null;
  }

  /**
   * Find the config within the root when in a workspace.
   */
  // eslint-disable-next-line complexity
  findConfigInWorkspaceRoot(root: string): ConfigPathOrObject | null {
    let currentDir = path.dirname(root);

    if (currentDir.includes('node_modules')) {
      return null;
    }

    this.debug('Detecting if in a workspace');

    let workspaceRoot = '';
    let workspacePackage: any = {};
    let workspacePatterns: string[] = [];

    // eslint-disable-next-line no-constant-condition
    while (true) {
      if (!currentDir || currentDir === '.' || currentDir === '/') {
        break;
      }

      const pkgPath = path.join(currentDir, 'package.json');

      if (fs.existsSync(pkgPath)) {
        workspacePackage = this.parseFile(pkgPath);
      }

      workspaceRoot = currentDir;
      workspacePatterns = this.tool.getWorkspacePaths({
        relative: true,
        root: currentDir,
      });

      if (workspacePackage && workspacePatterns.length > 0) {
        break;
      }

      currentDir = path.dirname(currentDir);
    }

    if (!workspaceRoot) {
      this.debug('No workspace found');

      return null;
    }

    const match = workspacePatterns.some(
      (pattern: string) => !!root.match(new RegExp(path.join(workspaceRoot, pattern), 'u')),
    );

    this.debug.invariant(
      match,
      `Matching patterns: ${workspacePatterns.map(p => chalk.cyan(p)).join(', ')}`,
      'Match found',
      'Invalid workspace package',
    );

    if (!match) {
      return null;
    }

    this.workspaceRoot = workspaceRoot;

    return (
      this.findConfigInPackageJSON(workspacePackage) ||
      this.findConfigInLocalFiles(workspaceRoot) ||
      null
    );
  }

  /**
   * Return the config name used for file names and the package.json property.
   */
  getConfigName(): string {
    const { configName, appName } = this.tool.options;

    return camelCase(configName || appName);
  }

  /**
   * Inherit configuration settings from defined CLI options.
   */
  inheritFromArgs<T>(config: T, args: Arguments): T {
    const nextConfig: any = { ...config };
    const pluginTypes = this.tool.getRegisteredPlugins();
    const keys = new Set([
      'debug',
      'locale',
      'output',
      'silent',
      'theme',
      ...Object.keys(this.tool.options.configBlueprint),
    ]);

    this.debug('Inheriting config from CLI options');

    Object.keys(args).forEach(key => {
      const value = args[key];

      if (key === 'config' || key === 'extends' || key === 'settings') {
        return;
      }

      // @ts-ignore Ignore symbol check
      const pluginType = pluginTypes[key as keyof typeof pluginTypes];

      // Plugins
      if (pluginType) {
        const { pluralName, singularName } = pluginType;

        this.debug('  --%s=[%s]', singularName, value.join(', '));
        nextConfig[pluralName] = (nextConfig[pluralName] || []).concat(value);

        // Other
      } else if (keys.has(key)) {
        this.debug('  --%s=%s', key, value);
        nextConfig[key] = value;
      }
    });

    return nextConfig;
  }

  /**
   * Load a local configuration file relative to the current working directory,
   * or from within a package.json property of the same appName.
   *
   * Support both JSON and JS file formats by globbing the config directory.
   */
  loadConfig<T>(args: Arguments): T {
    if (isEmptyObject(this.package) || !this.package.name) {
      throw new Error(this.tool.msg('errors:packageJsonNotLoaded'));
    }

    this.debug('Locating configuration');

    const { configBlueprint, settingsBlueprint, root } = this.tool.options;
    const pluginsBlueprint: any = {};
    const configPath =
      this.findConfigFromArg(args.config) ||
      this.findConfigInPackageJSON(this.package) ||
      this.findConfigInLocalFiles(root) ||
      this.findConfigInWorkspaceRoot(root);

    if (!configPath) {
      throw new Error(this.tool.msg('errors:configNotFound'));
    }

    Object.values(this.tool.getRegisteredPlugins()).forEach(type => {
      const { contract, singularName, pluralName } = type!;

      this.debug('Generating %s blueprint', chalk.magenta(singularName));

      // prettier-ignore
      pluginsBlueprint[pluralName] = array(union<PluginSetting<Function>>([
        string().notEmpty(),
        shape({ [singularName]: string().notEmpty() }),
        instance(contract, true),
      ], []));
    });

    const config = optimal(
      this.inheritFromArgs(this.parseAndExtend(configPath), args),
      {
        ...configBlueprint,
        ...pluginsBlueprint,
        debug: bool(),
        extends: array(string()),
        locale: string(),
        output: number(2).between(1, 3, true),
        // shape() requires a non-empty object
        settings: isEmptyObject(settingsBlueprint) ? object() : shape(settingsBlueprint),
        silent: bool(),
        theme: string('default').notEmpty(),
      },
      {
        file: typeof configPath === 'string' ? path.basename(configPath) : '',
        name: 'ConfigLoader',
        unknown: true,
      },
    );

    return config as T;
  }

  /**
   * Load the "package.json" from the current working directory,
   * as we require the dev tool to be ran from the project root.
   */
  loadPackageJSON(): PackageConfig {
    const { root } = this.tool.options;
    const filePath = path.join(root, 'package.json');

    this.debug('Locating package.json in %s', chalk.cyan(root));

    if (!fs.existsSync(filePath)) {
      throw new Error(this.tool.msg('errors:packageJsonNotFound'));
    }

    this.package = optimal(
      this.parseFile(filePath),
      {
        name: string().notEmpty(),
        version: string('0.0.0'),
      },
      {
        file: 'package.json',
        name: 'ConfigLoader',
        unknown: true,
      },
    );

    return this.package;
  }

  /**
   * If an `extends` option exists, recursively merge the current configuration
   * with the preset configurations defined within `extends`,
   * and return the new configuration object.
   */
  parseAndExtend(fileOrConfig: ConfigPathOrObject): ConfigObject {
    let config;
    let baseDir = '';

    // Parse out the object if a file path
    if (typeof fileOrConfig === 'string') {
      config = this.parseFile(fileOrConfig);
      baseDir = path.dirname(fileOrConfig);
    } else {
      config = fileOrConfig;
    }

    // Verify we're working with an object
    if (!isObject(config)) {
      throw new Error(this.tool.msg('errors:configInvalid'));
    }

    const { extends: extendPaths } = config;

    // Nothing to extend, so return the current config
    if (!extendPaths || extendPaths.length === 0) {
      return config;
    }

    // Resolve extend paths and inherit their config
    const nextConfig = {};
    const resolvedPaths = this.resolveExtendPaths(extendPaths, baseDir);

    resolvedPaths.forEach(extendPath => {
      if (this.parsedFiles[extendPath]) {
        return;
      }

      if (!fs.existsSync(extendPath)) {
        throw new Error(this.tool.msg('errors:presetConfigNotFound', { extendPath }));
      } else if (!fs.statSync(extendPath).isFile()) {
        throw new Error(this.tool.msg('errors:presetConfigInvalid', { extendPath }));
      }

      this.debug('Extending from file %s', chalk.cyan(extendPath));

      mergeWith(nextConfig, this.parseAndExtend(extendPath), handleMerge);
    });

    // Apply the current config after extending preset configs
    config.extends = resolvedPaths;

    mergeWith(nextConfig, config, handleMerge);

    return nextConfig;
  }

  /**
   * Parse a configuration file at the defined file system path.
   * If the file ends in "json" or "json5", parse it with JSON5.
   * If the file ends in "js", import the file and use the default object.
   * Otherwise throw an error.
   */
  parseFile(filePath: string, args: any[] = [], options: ParseOptions = {}): ConfigObject {
    const name = path.basename(filePath);
    const ext = path.extname(filePath);
    let value: any = null;

    this.debug('Parsing file %s', chalk.cyan(filePath));

    if (!path.isAbsolute(filePath)) {
      throw new Error(this.tool.msg('errors:absolutePathRequired'));
    }

    if (ext === '.json' || ext === '.json5') {
      value = JSON5.parse(fs.readFileSync(filePath, 'utf8'));
    } else if (ext === '.js') {
      value = requireModule(filePath);

      if (typeof value === 'function') {
        if (options.errorOnFunction) {
          throw new Error(this.tool.msg('errors:configNoFunction', { name }));
        } else {
          value = value(...args);
        }
      }
    } else {
      throw new Error(this.tool.msg('errors:configUnsupportedExt', { ext }));
    }

    if (!isObject(value)) {
      throw new Error(this.tool.msg('errors:configInvalidNamed', { name }));
    }

    this.parsedFiles[filePath] = true;

    return value;
  }

  /**
   * Resolve file system paths for the `extends` configuration value
   * using the following guidelines:
   *
   *  - Absolute paths should be normalized and used as is.
   *  - Relative paths should be resolved relative to the CWD.
   *  - Strings that match a node module name should resolve to a config file relative to the CWD.
   *  - Strings that start with "<plugin>:" should adhere to the previous rule.
   */
  resolveExtendPaths(extendPaths: string[], baseDir: string = ''): string[] {
    return extendPaths.map(extendPath => {
      if (typeof extendPath !== 'string') {
        throw new TypeError(this.tool.msg('errors:configExtendsInvalid'));
      }

      const { appName, scoped, root } = this.tool.options;
      const configName = this.getConfigName();
      let match = null;

      // Absolute path, use it directly
      if (path.isAbsolute(extendPath)) {
        return path.normalize(extendPath);

        // Relative path, resolve with parent folder or cwd
      } else if (extendPath[0] === '.') {
        return path.resolve(baseDir || root, extendPath);

        // Node module, resolve to a config file
      } else if (extendPath.match(MODULE_NAME_PATTERN)) {
        return this.resolveModuleConfigPath(configName, extendPath, true);

        // Plugin, resolve to a node module
      } else if ((match = extendPath.match(PLUGIN_NAME_PATTERN))) {
        return this.resolveModuleConfigPath(
          configName,
          formatModuleName(appName, match[1], extendPath, scoped),
          true,
        );
      }

      throw new Error(this.tool.msg('errors:configExtendsInvalidPath', { extendPath }));
    });
  }

  /**
   * Resolve a Node/NPM module path to an app config file.
   */
  resolveModuleConfigPath(
    configName: string,
    moduleName: string,
    preset: boolean = false,
    ext: string = 'js',
  ): string {
    const fileName = preset ? `${configName}.preset.${ext}` : `${configName}.${ext}`;

    return path.resolve(this.tool.options.root, 'node_modules', moduleName, 'configs', fileName);
  }
}
