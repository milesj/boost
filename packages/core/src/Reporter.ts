/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import chalk from 'chalk';
import Console from './Console';
import Module from './Module';
import ModuleLoader from './ModuleLoader';
import Task from './Task';
import themePalettes from './themes';
import { Color, ColorType, ColorPalette } from './types';

export const SLOW_THRESHOLD = 10000; // ms

export default class Reporter<Line = string, Options = {}> extends Module<Options> {
  // @ts-ignore Set after instantiation
  console: Console;

  lines: Line[] = [];

  startTime: number = 0;

  stopTime: number = 0;

  /**
   * Register console listeners.
   */
  bootstrap() {
    this.console.on('start', this.handleBaseStart).on('stop', this.handleBaseStop);
  }

  /**
   * Add a line to be rendered.
   */
  addLine(line: Line): this {
    this.lines.push(line);

    return this;
  }

  /**
   * Display an error and it's stack.
   */
  displayError(error: Error): void {
    this.console.write(`\n${this.style(error.message, 'failure', ['bold'])}\n`);

    // Remove message line from stack
    if (error.stack) {
      const stack = this.style(
        error.stack
          .split('\n')
          .slice(1)
          .join('\n')
          .trim(),
        'pending',
      );

      this.console.write(stack, 1);
    }

    this.console.write('\n');
  }

  /**
   * Find a line using a callback
   */
  findLine(callback: (item: Line) => boolean): Line | undefined {
    return this.lines.find(line => callback(line));
  }

  /**
   * Return specific colors based on chosen theme.
   */
  getColorPalette(): ColorPalette {
    const { theme } = this.console.options;
    let palette = {};

    if (chalk.level >= 2) {
      if (themePalettes[theme]) {
        palette = themePalettes[theme];
      } else if (theme !== 'default') {
        palette = new ModuleLoader<ColorPalette>(this.console.tool, 'theme', null, true).loadModule(
          theme,
        );
      }
    }

    return {
      default: 'white',
      failure: 'red',
      pending: 'gray',
      success: 'green',
      warning: 'yellow',
      ...palette,
    };
  }

  /**
   * Return a specific color for each task status.
   */
  getColorType(task: Task<any>): ColorType {
    if (task.isSkipped()) {
      return 'warning';
    } else if (task.hasPassed()) {
      return 'success';
    } else if (task.hasFailed()) {
      return 'failure';
    }

    return 'pending';
  }

  /**
   * Calculate the elapsed time and highlight as red if over the threshold.
   */
  getElapsedTime(start: number, stop: number = 0, highlight: boolean = true): string {
    const time = (stop || Date.now()) - start;
    const isSlow = time > SLOW_THRESHOLD;
    const elapsed = `${(time / 1000).toFixed(2)}s`; // eslint-disable-line no-magic-numbers

    return isSlow && highlight ? this.style(elapsed, 'failure') : elapsed;
  }

  /**
   * Set start time.
   */
  handleBaseStart = () => {
    this.startTime = Date.now();
  };

  /**
   * Set stop time and render.
   */
  handleBaseStop = () => {
    this.stopTime = Date.now();
  };

  /**
   * Create an indentation based on the defined length.
   */
  indent(length: number = 0): string {
    return ' '.repeat(length);
  }

  /**
   * Remove a line to be rendered.
   */
  removeLine(callback: (line: Line) => boolean): this {
    this.lines = this.lines.filter(line => !callback(line));

    return this;
  }

  /**
   * Create a chalk formatted string with accessible colors and modifiers applied.
   */
  style(
    message: string,
    type: ColorType = 'default',
    modifiers: ('reset' | 'bold' | 'dim' | 'italic' | 'underline' | 'inverse' | 'hidden')[] = [],
  ): string {
    const color = this.getColorPalette()[type];
    let out = color.charAt(0) === '#' ? chalk.hex(color) : chalk[color as Color];

    modifiers.forEach(modifier => {
      out = out[modifier];
    });

    return out(message);
  }
}
