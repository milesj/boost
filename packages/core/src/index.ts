/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import CLI from './CLI';
import ConfigLoader from './ConfigLoader';
import Console from './Console';
import Context from './Context';
import Emitter, { EventArguments, EventListener } from './Emitter';
import ModuleLoader from './ModuleLoader';
import Pipeline from './Pipeline';
import Plugin from './Plugin';
import Reporter from './Reporter';
import Routine from './Routine';
import Task from './Task';
import Tool, { ToolOptions, ToolConfig, ToolPluginRegistry } from './Tool';

export {
  CLI,
  ConfigLoader,
  Console,
  Context,
  Emitter,
  EventArguments,
  EventListener,
  ModuleLoader,
  Pipeline,
  Plugin,
  Reporter,
  Routine,
  Task,
  Tool,
  ToolConfig,
  ToolOptions,
  ToolPluginRegistry,
};

export * from './types';
