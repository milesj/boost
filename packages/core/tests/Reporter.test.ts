import chalk from 'chalk';
import Reporter from '../src/Reporter';
import Task from '../src/Task';
import { STATUS_PASSED, STATUS_FAILED } from '../src/constants';
import { createTestConsole } from './helpers';

describe('Reporter', () => {
  let reporter: Reporter<any, any>;

  beforeEach(() => {
    reporter = new Reporter();
    reporter.console = createTestConsole();
  });

  describe('bootstrap()', () => {
    it('sets start and stop events', () => {
      const spy = jest.spyOn(reporter.console, 'on');

      reporter.bootstrap();

      expect(spy).toHaveBeenCalledWith('start', expect.anything());
      expect(spy).toHaveBeenCalledWith('stop', expect.anything());
    });
  });

  describe('addLine()', () => {
    it('adds a line to the list', () => {
      expect(reporter.lines).toEqual([]);

      reporter.addLine('foo');

      expect(reporter.lines).toEqual(['foo']);
    });
  });

  describe('displayError()', () => {
    it('writes to stderr', () => {
      reporter.displayError(new Error('Oops'));

      expect(reporter.console.write).toHaveBeenCalledTimes(3);
    });
  });

  describe('findLine()', () => {
    it('returns undefined if not found', () => {
      expect(reporter.findLine(line => line === 'foo')).toBeUndefined();
    });

    it('returns the line', () => {
      reporter.lines.push('foo');

      expect(reporter.findLine(line => line === 'foo')).toBe('foo');
    });
  });

  describe('getColorPalette()', () => {
    const oldLevel = chalk.level;
    const basePalette = {
      failure: 'red',
      pending: 'gray',
      success: 'green',
      warning: 'yellow',
    };

    afterEach(() => {
      chalk.level = oldLevel;
    });

    it('returns base palette if chalk level < 2', () => {
      chalk.level = 1;

      expect(reporter.getColorPalette()).toEqual(basePalette);
    });

    it('returns base palette if chalk level >= 2 and theme is default', () => {
      chalk.level = 2;
      reporter.console.options.theme = 'default';

      expect(reporter.getColorPalette()).toEqual(basePalette);
    });

    it('returns base palette if theme does not exist', () => {
      chalk.level = 2;
      reporter.console.options.theme = 'unknown';

      expect(reporter.getColorPalette()).toEqual(basePalette);
    });

    it('returns theme palette', () => {
      chalk.level = 2;
      reporter.console.options.theme = 'solarized';

      expect(reporter.getColorPalette()).toEqual({
        failure: '#dc322f',
        pending: '#93a1a1',
        success: '#859900',
        warning: '#b58900',
      });
    });
  });

  describe('getColorType()', () => {
    it('returns yellow for skipped', () => {
      const task = new Task('task').skip();

      expect(reporter.getColorType(task)).toBe('warning');
    });

    it('returns green for passed', () => {
      const task = new Task('task');

      task.status = STATUS_PASSED;

      expect(reporter.getColorType(task)).toBe('success');
    });

    it('returns red for failed', () => {
      const task = new Task('task');

      task.status = STATUS_FAILED;

      expect(reporter.getColorType(task)).toBe('failure');
    });

    it('returns gray otherwise', () => {
      const task = new Task('task', () => {});

      expect(reporter.getColorType(task)).toBe('pending');
    });
  });

  describe('getElapsedTime()', () => {
    it('returns numbers in seconds', () => {
      expect(reporter.getElapsedTime(1000, 5000)).toBe('4.00s');
    });

    it('colors red if higher than slow threshold', () => {
      expect(reporter.getElapsedTime(1000, 15000)).toBe(chalk.red('14.00s'));
    });

    it('doesnt color if highlight is false', () => {
      expect(reporter.getElapsedTime(1000, 15000, false)).toBe('14.00s');
    });
  });

  describe('handleBaseStart()', () => {
    const oldCI = process.env.CI;

    beforeEach(() => {
      process.env.CI = '';
      jest.useFakeTimers();
    });

    afterEach(() => {
      jest.useRealTimers();
      process.env.CI = oldCI;
    });

    it('sets start time', () => {
      reporter.handleBaseStart();

      expect(reporter.startTime).not.toBe(0);
    });
  });

  describe('handleBaseStop()', () => {
    it('sets stop time', () => {
      reporter.handleBaseStop();

      expect(reporter.stopTime).not.toBe(0);
    });
  });

  describe('indent()', () => {
    it('indents based on length', () => {
      expect(reporter.indent()).toBe('');
      expect(reporter.indent(1)).toBe(' ');
      expect(reporter.indent(3)).toBe('   ');
    });
  });

  describe('removeLine()', () => {
    it('removes a line', () => {
      reporter.lines.push('foo', 'bar', 'baz');
      reporter.removeLine(line => line === 'foo');

      expect(reporter.lines).toEqual(['bar', 'baz']);
    });
  });

  describe('style()', () => {
    it('colors pending', () => {
      expect(reporter.style('foo', 'pending')).toBe(chalk.gray('foo'));
    });

    it('colors failure', () => {
      expect(reporter.style('foo', 'failure')).toBe(chalk.red('foo'));
    });

    it('colors success', () => {
      expect(reporter.style('foo', 'success')).toBe(chalk.green('foo'));
    });

    it('colors warning', () => {
      expect(reporter.style('foo', 'warning')).toBe(chalk.yellow('foo'));
    });

    it('can apply modifiers', () => {
      expect(reporter.style('foo', 'pending', ['bold', 'dim', 'italic'])).toBe(
        chalk.gray.bold.dim.italic('foo'),
      );
    });
  });
});
