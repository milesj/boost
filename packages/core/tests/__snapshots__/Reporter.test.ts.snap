// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reporter getColorPalette() errors if theme does not exist 1`] = `"Missing theme. Attempted import in order: test-boost-theme-custom, @boost/theme-custom, boost-theme-custom"`;
